src/Evolution/Systems/Cce/InterfaceManagers/GhLocalTimeStepping.hpp

```diff
@@ -26,75 +27,64 @@ namespace Cce::InterfaceManagers {
+ * current `TimeStepId` via `GhLocalTimeStepping::insert_gh_data()`. The CCE
+ * system supplies requests for time steps via
+ * `GhLocalTimeStepping::request_gh_data()` and receives interpolated boundary
+ * data via `GhLocalTimeStepping::retrieve_and_remove_first_ready_gh_data()`.
  */
 class GhLocalTimeStepping : public GhInterfaceManager {
  public:
-  using dt_gh_variables = Variables<tmpl::list<
-      ::Tags::dt<gr::Tags::SpacetimeMetric<3, ::Frame::Inertial, DataVector>>,
-      ::Tags::dt<GeneralizedHarmonic::Tags::Pi<3, ::Frame::Inertial>>,
-      ::Tags::dt<GeneralizedHarmonic::Tags::Phi<3, ::Frame::Inertial>>>>;
-
-  struct AdamsBashforthOrder {
-    using type = size_t;
+  struct BoundaryInterpolator {
+    using type = std::unique_ptr<intrp::SpanInterpolator>;
     static constexpr Options::String help = {
-        "Convergence order for the internal Adams-Bashforth stepper"};
-    static type lower_bound() noexcept { return 1; }
-    static type upper_bound() noexcept {
-      return TimeSteppers::AdamsBashforthN::maximum_order;
-    }
+        "Interpolator for computing CCE data from GH time points"};
   };
 
   static constexpr Options::String help{
-      "Pass data between GH and CCE systems via Adams-Bashforth local "
-      "time-stepping"};
+      "Interpolate data from the GH system to generate CCE inputs"};
 
-  using options = tmpl::list<AdamsBashforthOrder>;
+  using options = tmpl::list<BoundaryInterpolator>;
 
   GhLocalTimeStepping() = default;
+  GhLocalTimeStepping(const GhLocalTimeStepping& rhs) noexcept
+      : gh_data_{rhs.gh_data_}, requests_{rhs.requests_} {
+    if (rhs.interpolator_.get() != nullptr) {
+      interpolator_ = rhs.interpolator_->get_clone();
+    }
+  }
 
-  explicit GhLocalTimeStepping(const size_t order)
-      : order_{order}, boundary_history_{1_st}, time_stepper_{order} {}
+  explicit GhLocalTimeStepping(
+      std::unique_ptr<intrp::SpanInterpolator> interpolator)
+      : interpolator_{std::move(interpolator)} {}
+  GhLocalTimeStepping& operator=(const GhLocalTimeStepping& rhs) noexcept {
+    if (rhs.interpolator_.get() != nullptr) {
+      interpolator_ = rhs.interpolator_->get_clone();
+    }
+    gh_data_ = rhs.gh_data_;
+    requests_ = rhs.requests_;
+    return *this;
+  }
 
   explicit GhLocalTimeStepping(CkMigrateMessage* /*unused*/) noexcept {}
 
   WRAPPED_PUPable_decl_template(GhLocalTimeStepping);  // NOLINT
 
   std::unique_ptr<GhInterfaceManager> get_clone() const noexcept override;
 
-  /// \brief Store the provided data set to prepare for time-stepping dense
-  /// output.
-  ///
-  /// \details The `next_time_id` is required to infer the span of time values
-  /// that should be permitted for the dense output, and at what point the CCE
-  /// system should wait for additional data from the GH system.
-  void insert_gh_data(TimeStepId time_id,
+  /// \brief Store the provided data set to prepare for interpolation.
+  void insert_gh_data(double time,
```

Comment:
The way I have it working is that there is an action that inserts requests for the next CCE time, and an action that inserts new data from GH interpolation. Each time there is either a new request or new gh data it checks the `retrieve_and_remove_first_ready_gh_data`, and sends it to the CCE evolution component if its ready, so I don't think there's any trouble with sequencing between the components.
The GH data is stored in a map, so I think there shouldn't be any trouble with misordered arrival of the GH data.
The CCE requests are stored in a `std::deque`, so there is an assumption here that the CCE requests are ordered, but I think that is guaranteed by the communication pattern in CCE -- it always consumes a time before requesting the next. For flexibility, though, I'll to change it to a `std::set`, then there aren't any ordering constraints for these structures.

Reply:
I think that answers my question, but I'm going to ask another clarifying one to be sure.

The thing I'm worried about is that the `create_span_for_time_value` function makes decisions based on what GH data is available.  Is the set of available data deterministic when that function is called?

Reply:
Ah, right, that's a good point. That does make ordering assumptions -- In particular, one could imagine that if the data was inserted such that most of the data in a part of the evolution arrives in sequence, but one intermediate value arrives far later, then this interpolation would activate using more distant points, and presumably have degraded accuracy because of the missing point.

I think I can solve this if I can send along with each set of data from a dense trigger, the previous time the dense trigger activated -- then the interpolation step in this class could check that the set of data it is using to perform the interpolation is contiguous.
Is there an easy way to get access to the previous trigger time so that I could include that information?

Reply:
You can change `DenseTrigger::is_triggered` to non-const and use a member variable.  I was planning to do that anyway because the control system trigger has some similar needs.